daemon/cmd/ipcache.go

```diff
@@ -87,3 +105,144 @@ func containsSubnet(outer, inner net.IPNet) bool {
+	nidsToReserve := make([]identity.NumericIdentity, 0, len(d.restoredCIDRs))
+
+	for prefix, nid := range d.restoredCIDRs {
+		nidsToReserve = append(nidsToReserve, nid)
+		metaUpdates = append(metaUpdates, ipcache.MU{
+			Prefix:   prefix,
+			Source:   source.Restored,
+			Resource: restoredCIDRResource,
+			Metadata: []ipcache.IPMetadata{ipcachetypes.RequestedIdentity(nid)},
+		})
+	}
+	for _, prefix := range ingressIPs {
+		metaUpdates = append(metaUpdates, ipcache.MU{
+			Prefix:   prefix,
+			Source:   source.Restored,
+			Resource: ingressResource,
+			Metadata: []ipcache.IPMetadata{labels.LabelIngress},
+		})
+
+		// Set any restored ingress IPs back on the LocalNode object
+		d.nodeLocalStore.Update(func(n *node.LocalNode) {
+			addr := prefix.Addr()
+			if addr.Is4() {
+				n.IPv4IngressIP = addr.AsSlice()
+			} else {
+				n.IPv6IngressIP = addr.AsSlice()
+			}
+		})
+	}
+	if len(ingressIPs) > 0 {
+		log.WithField(logfields.Ingress, ingressIPs).Info("Restored ingress IPs")
+	}
+
+	// Insert the batched changes in to the ipcache.
+	// Even though the ipcache map hasn't been initialized yet, this is
+	// safe to do so, because the ipcache's apply controller is currently
+	// paused.
+	d.ipcache.IdentityAllocator.ReserveLocalIdentities(nidsToReserve)
+	d.ipcache.UpsertMetadataBatch(metaUpdates...)
+
+	return nil
+}
+
+// releaseRestoredCIDRS removes the placeholder metadata that was inserted
+// in to the ipcache when local identities were restored.
+// Any identities actually in use will still exist after this.
+//
+// This should be called after a grace period (default 10 minutes, set
+// by --identity-restore-grace-period).
+// This grace period is needed when running on an external workload
+// where policy synchronization is not done via k8s. Also in k8s
+// case it is prudent to allow concurrent endpoint regenerations to
+// (re-)allocate the restored identities before we release them.
+//
+// Any CIDRs still in use after the grace period will have other sources
+// of metadata in the ipcache, and thus will remain. CIDRs for which
+// restoration was the only source of metadata will be deallocated.
+func (d *Daemon) releaseRestoredCIDRs() {
+	defer func() {
+		// release the memory held by restored CIDRs
+		d.restoredCIDRs = nil
+	}()
+	if len(d.restoredCIDRs) == 0 {
+		return
+	}
+
+	log.WithField(logfields.Count, len(d.restoredCIDRs)).Info("Removing identity reservations for restored CIDR identities")
+	updates := make([]ipcache.MU, 0, len(d.restoredCIDRs))
+	nids := make([]identity.NumericIdentity, 0, len(d.restoredCIDRs))
+	for prefix, nid := range d.restoredCIDRs {
+		nids = append(nids, nid)
+		updates = append(updates, ipcache.MU{
+			Prefix:   prefix,
+			Resource: restoredCIDRResource,
+			Metadata: []ipcache.IPMetadata{ipcachetypes.RequestedIdentity(0)},
+		})
+	}
+
+	d.ipcache.RemoveMetadataBatch(updates...)
+	d.ipcache.UnreserveLocalIdentities(nids)
```

Comment:
nit: The way I see the reserve / unreserve being tied to the metadata batch, it makes me wonder whether the ipcache should just make the reservation when inserting metadata with this type. One reason I thought of this is that it seems like the whole reason we allocate these `nids` on the stack is in order to have a slightly-differently-typed argument and I wondered whether these two could share the same argument, then wondered if they could just be the same call and simplify usage from `daemon`. 

But then I recognize that lower down we probably need to create the slice and do the datatype swap anyway, so we're probably just shuffling the logic around slightly if we did something like this. We could always iterate and tidy and improve if we think there's a better way on this, so no need to address at this time. Feel free to resolve this comment without action.

Reply:
I did think about that; it felt funny to touch the identity allocator in the "entrypoint" methods. The precedent is that the id allocator is only touched in `resolveIdentity()`. Not that we can't override precedent, mind.

I *hope* nothing else uses the identity reservation system :-).